File: internal/cluster/cluster.go
package cluster

import (
	"context"
	"encoding/json"
	"fmt"
	"time"

	"github.com/redis/go-redis/v9"
)

// Cluster provides inter-broker routing via Redis PubSub.

type Cluster struct {
	nodeID string
	cli    redis.UniversalClient
	prefix string
	onMsg  func(msg *ClusterMessage)
	cancel context.CancelFunc
}

type ClusterMessage struct {
	From    string `json:"from"`
	Topic   string `json:"topic"`
	Payload []byte `json:"payload"`
	QoS     byte   `json:"qos"`
	Retain  bool   `json:"retain"`
}

func New(cli redis.UniversalClient, nodeID, prefix string, onMsg func(*ClusterMessage)) *Cluster {
	if prefix == "" {
		prefix = "mqtt"
	}
	return &Cluster{nodeID: nodeID, cli: cli, prefix: prefix, onMsg: onMsg}
}

func (c *Cluster) Start(ctx context.Context) error {
	ctx2, cancel := context.WithCancel(ctx)
	c.cancel = cancel

	// node heartbeat
	go c.heartbeatLoop(ctx2)

	channel := c.prefix + ":cluster"
	pubsub := c.cli.Subscribe(ctx2, channel)
	if _, err := pubsub.Receive(ctx2); err != nil {
		return err
	}
	go func() {
		defer pubsub.Close()
		ch := pubsub.Channel()
		for {
			select {
			case <-ctx2.Done():
				return
			case m, ok := <-ch:
				if !ok {
					return
				}
				var cm ClusterMessage
				if err := json.Unmarshal([]byte(m.Payload), &cm); err != nil {
					continue
				}
				if cm.From == c.nodeID {
					continue
				}
				if c.onMsg != nil {
					c.onMsg(&cm)
				}
			}
		}
	}()
	return nil
}

func (c *Cluster) Stop() {
	if c.cancel != nil {
		c.cancel()
	}
	// remove node key
	_ = c.cli.Del(context.Background(), c.prefix+":nodes:"+c.nodeID).Err()
}

func (c *Cluster) heartbeatLoop(ctx context.Context) {
	key := c.prefix + ":nodes:" + c.nodeID
	ticker := time.NewTicker(5 * time.Second)
	defer ticker.Stop()
	// initial register
	_ = c.cli.Set(ctx, key, time.Now().Unix(), 15*time.Second).Err()
	for {
		select {
		case <-ctx.Done():
			return
		case <-ticker.C:
			_ = c.cli.Set(ctx, key, time.Now().Unix(), 15*time.Second).Err()
		}
	}
}

func (c *Cluster) Publish(ctx context.Context, topic string, payload []byte, qos byte, retain bool) error {
	cm := ClusterMessage{From: c.nodeID, Topic: topic, Payload: payload, QoS: qos, Retain: retain}
	data, err := json.Marshal(cm)
	if err != nil {
		return err
	}
	return c.cli.Publish(ctx, c.prefix+":cluster", data).Err()
}

func (c *Cluster) Nodes(ctx context.Context) ([]string, error) {
	var out []string
	iter := c.cli.Scan(ctx, 0, c.prefix+":nodes:*", 0).Iterator()
	for iter.Next(ctx) {
		k := iter.Val()
		// k = prefix:nodes:nodeID
		// extract nodeID
		p := c.prefix + ":nodes:"
		if len(k) > len(p) {
			out = append(out, k[len(p):])
		}
	}
	return out, nil
}

var _ = fmt.Sprintf

File: src/cluster.rs
//! Inter-broker routing via Redis PubSub. Port of `internal/cluster/cluster.go`.
//!
//! - heartbeat: `SET mqtt:nodes:<id> <unix> EX 15s` every 5s
//! - message bus: `PUBLISH mqtt:cluster` (JSON ClusterMessage)
//! - meta bus: `PUBLISH mqtt:cluster:meta` (JSON ClusterMeta, sub/unsub)

use futures_util::StreamExt;
use redis::aio::ConnectionManager;
use serde::{Deserialize, Serialize};
use tokio::sync::mpsc;

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ClusterMessage {
    #[serde(rename = "from")]
    pub from: String,
    #[serde(rename = "topic")]
    pub topic: String,
    #[serde(rename = "payload", with = "crate::session::serde_bytes_b64")]
    pub payload: Vec<u8>,
    #[serde(rename = "qos")]
    pub qos: u8,
    #[serde(rename = "retain")]
    pub retain: bool,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ClusterMeta {
    #[serde(rename = "from")]
    pub from: String,
    #[serde(rename = "action")]
    pub action: String, // sub / unsub
    #[serde(rename = "filter")]
    pub filter: String,
}

pub struct Cluster {
    node_id: String,
    client: redis::Client,
    conn: ConnectionManager,
    prefix: String,
}

impl Cluster {
    pub fn new(client: redis::Client, conn: ConnectionManager, node_id: &str, prefix: &str) -> Self {
        Self {
            node_id: node_id.to_string(),
            client,
            conn,
            prefix: if prefix.is_empty() { "mqtt".into() } else { prefix.to_string() },
        }
    }

    pub fn node_id(&self) -> &str {
        &self.node_id
    }

    /// Start heartbeat + pubsub listener. Returns a shutdown handle.
    pub async fn start(
        &self,
        on_msg: mpsc::Sender<ClusterMessage>,
        on_meta: mpsc::Sender<ClusterMeta>,
    ) -> Result<ClusterHandle, String> {
        let channel = format!("{}:cluster", self.prefix);
        let meta_channel = format!("{}:cluster:meta", self.prefix);

        // dedicated pubsub connection
        let mut pubsub = self.client.get_async_pubsub().await.map_err(|e| e.to_string())?;
        pubsub.subscribe(&channel).await.map_err(|e| e.to_string())?;
        pubsub.subscribe(&meta_channel).await.map_err(|e| e.to_string())?;

        // heartbeat task
        let hb = {
            let conn = self.conn.clone();
            let key = format!("{}:nodes:{}", self.prefix, self.node_id);
            tokio::spawn(async move {
                let mut conn = conn;
                let _: Result<(), _> = redis::cmd("SET")
                    .arg(&key)
                    .arg(crate::auth::now_unix() as i64)
                    .arg("EX")
                    .arg(15usize)
                    .query_async(&mut conn)
                    .await;
                let mut ticker = tokio::time::interval(std::time::Duration::from_secs(5));
                ticker.set_missed_tick_behavior(tokio::time::MissedTickBehavior::Delay);
                loop {
                    ticker.tick().await;
                    let _: Result<(), _> = redis::cmd("SET")
                        .arg(&key)
                        .arg(crate::auth::now_unix() as i64)
                        .arg("EX")
                        .arg(15usize)
                        .query_async(&mut conn)
                        .await;
                }
            })
        };

        // pubsub dispatch task
        let node_id = self.node_id.clone();
        let dispatch = tokio::spawn(async move {
            let mut stream = pubsub.into_on_message();
            while let Some(msg) = stream.next().await {
                let ch = msg.get_channel_name().to_string();
                let payload: String = msg.get_payload().unwrap_or_default();
                if ch == meta_channel_ref(&meta_channel) || ch == meta_channel {
                    if let Ok(meta) = serde_json::from_str::<ClusterMeta>(&payload) {
                        if meta.from == node_id {
                            continue;
                        }
                        if on_meta.send(meta).await.is_err() {
                            break;
                        }
                    }
                    continue;
                }
                if let Ok(cm) = serde_json::from_str::<ClusterMessage>(&payload) {
                    if cm.from == node_id {
                        continue;
                    }
                    if on_msg.send(cm).await.is_err() {
                        break;
                    }
                }
            }
        });

        Ok(ClusterHandle { hb: Some(hb), dispatch: Some(dispatch), conn: self.conn.clone(), node_key: format!("{}:nodes:{}", self.prefix, self.node_id) })
    }

    /// Broadcast a message to the cluster.
    pub async fn publish(&self, topic: &str, payload: &[u8], qos: u8, retain: bool) -> Result<(), String> {
        let cm = ClusterMessage {
            from: self.node_id.clone(),
            topic: topic.to_string(),
            payload: payload.to_vec(),
            qos,
            retain,
        };
        let data = serde_json::to_string(&cm).map_err(|e| e.to_string())?;
        let mut conn = self.conn.clone();
        let _: () = redis::cmd("PUBLISH")
            .arg(format!("{}:cluster", self.prefix))
            .arg(data)
            .query_async(&mut conn)
            .await
            .map_err(|e| e.to_string())?;
        Ok(())
    }

    /// Broadcast subscription meta (sub/unsub).
    pub async fn publish_meta(&self, action: &str, filter: &str) -> Result<(), String> {
        let meta = ClusterMeta { from: self.node_id.clone(), action: action.to_string(), filter: filter.to_string() };
        let data = serde_json::to_string(&meta).map_err(|e| e.to_string())?;
        let mut conn = self.conn.clone();
        let _: () = redis::cmd("PUBLISH")
            .arg(format!("{}:cluster:meta", self.prefix))
            .arg(data)
            .query_async(&mut conn)
            .await
            .map_err(|e| e.to_string())?;
        Ok(())
    }

    /// List live nodes via SCAN on `mqtt:nodes:*`.
    pub async fn nodes(&self) -> Result<Vec<String>, String> {
        let mut conn = self.conn.clone();
        let pattern = format!("{}:nodes:*", self.prefix);
        let p = format!("{}:nodes:", self.prefix);
        let keys = scan_keys(&mut conn, &pattern).await?;
        Ok(keys.iter().filter_map(|k| k.strip_prefix(&p).map(|s| s.to_string())).collect())
    }
}

fn meta_channel_ref(s: &str) -> String {
    s.to_string()
}

async fn scan_keys(conn: &mut ConnectionManager, pattern: &str) -> Result<Vec<String>, String> {
    let mut out = Vec::new();
    let mut cursor: u64 = 0;
    loop {
        let (next, batch): (u64, Vec<String>) = redis::cmd("SCAN")
            .arg(cursor)
            .arg("MATCH")
            .arg(pattern)
            .arg("COUNT")
            .arg(100usize)
            .query_async(conn)
            .await
            .map_err(|e| e.to_string())?;
        out.extend(batch);
        cursor = next;
        if cursor == 0 {
            break;
        }
    }
    Ok(out)
}

/// Handle controlling cluster background tasks.
pub struct ClusterHandle {
    hb: Option<tokio::task::JoinHandle<()>>,
    dispatch: Option<tokio::task::JoinHandle<()>>,
    conn: ConnectionManager,
    node_key: String,
}

impl ClusterHandle {
    /// Stop tasks and remove our node key (TTL 15s would also expire it).
    pub async fn stop(mut self) {
        if let Some(h) = self.hb.take() {
            h.abort();
        }
        if let Some(h) = self.dispatch.take() {
            h.abort();
        }
        let mut conn = self.conn.clone();
        let _: Result<(), _> = redis::cmd("DEL").arg(&self.node_key).query_async(&mut conn).await;
    }
}
